controlplane/kubeadm/internal/controllers/upgrade.go

```diff
@@ -74,8 +75,16 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, errors.Wrap(err, "failed to update the kubernetes version in the kubeadm config map")
 	}
 
-	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil {
-		imageRepository := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository
+	// We intentionally only parse major/minor/patch so that the subsequent code
+	// also already applies to beta versions of new releases.
+	parsedVersionTolerant, err := version.ParseMajorMinorPatchTolerant(kcp.Spec.Version)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to parse kubernetes version %q", kcp.Spec.Version)
+	}
+	// Get the set imageRepository or the correct value if nothing is set and a migration is necessary.
+	imageRepository := internal.ImageRepositoryFromClusterConfig(kcp.Spec.KubeadmConfigSpec.ClusterConfiguration, parsedVersionTolerant)
+
+	if imageRepository != "" {
```

Comment:
I assume we never have to set the imageRepository to an empty string in the ConfigMap if it was set before?

Reply:
Hmm good point. We did never do before if `kcp.Spec.KubeadmConfigSpec.ClusterConfiguration` was nil.

So I think we should remove that if, however it seems to cause issues in at least a unit test (nil pointer in https://github.com/kubernetes-sigs/cluster-api/blob/7a533b2f1db794dc8f61ab1021d966051beccf87/controlplane/kubeadm/internal/controllers/scale_test.go#L157).

I would propose restoring the old behaviour by only checking `if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil` here and open an issue for fixing that we don't update the imageRepository if ClusterConfiguration is nil.

WDYT @sbueringer ?